[ValueTracking] Refine known bits for linear interpolation patterns

SavchenkoValeriy opened the pull request:
In this patch, we try to detect the lerp pattern: a * (b - c) + c * d
where a >= 0, b >= 0, c >= 0, d >= 0, and b >= c.

In that particular case, we can use the following chain of reasoning:

  a * (b - c) + c * d <= a' * (b - c) + a' * c = a' * b where a' = max(a, d)

Since that is true for arbitrary a, b, c and d within our constraints, we can
conclude that:

  max(a * (b - c) + c * d) <= max(max(a), max(d)) * max(b) = U

Considering that any result of the lerp would be less or equal to U, it would
have at least the number of leading 0s as in U.

While being quite a specific situation, it is fairly common in computer
graphics in the shape of alpha blending.

In conjunction with #165877, increases vectorization factor for lerp loops.

SavchenkoValeriy committed:
[KnownBits][NFC] Add tests for lerp pattern

SavchenkoValeriy committed:
[KnownBits] Refine known bits for lerp

fhahn reviewed (commented):
> Since that is true for arbitrary a, b, c and d within our constraints, we can
conclude that:
> 
> max(a * (b - c) + c * d) <= max(max(a), max(d)) * max(b) = U

Would it be possible to add an Alive2 proof for the conclusion?

SavchenkoValeriy commented:
> > Since that is true for arbitrary a, b, c and d within our constraints, we can
> > conclude that:
> > max(a * (b - c) + c * d) <= max(max(a), max(d)) * max(b) = U
> 
> Would it be possible to add an Alive2 proof for the conclusion?

That's a great point and I believe that such statements should be proven. However, I'd disagree that Alive2 is the best tool for this particular change.

Alive2 proves individual transformations and I will provide Alive2 proofs for the tests that I added. Since I added a more general known bits calculation that in theory can work well with one transformation but not the other, I wrote a z3 proof to show that inferred known bits are correct: https://gist.github.com/SavchenkoValeriy/a7d3e40868b9d512a70260ae63d8413b

SavchenkoValeriy commented:
Alive2 struggles (times out) with 32 bit vectors for this particular proof: https://alive2.llvm.org/ce/z/S-78k-
In my z3 proof, even 16 bit took quite some times.

SavchenkoValeriy commented on llvm/lib/Analysis/ValueTracking.cpp:
```diff
@@ -350,6 +350,140 @@ unsigned llvm::ComputeMaxSignificantBits(const Value *V, const DataLayout &DL,
+    // guaranteed to be non-negative as well.
+    return m_CombineOr(m_NUWSub(m_Value(B), m_Value(C)),
+                       m_Xor(m_Value(C), m_Value(B)));
+  };
```
While I agree with this sentiment about the patches in general, I believe that this patch is quite small (effectively ~50 LoC) and general. Ripping it apart to support individual sub-cases feels like going out of the way just to follow that principle.

For example, subtraction vs exclusive or will affect literally this line alone and remove just that line of code.

If you insist, of course I can do it, but I seriously don't see a good reason for that  especially considering that the proof works in a more general case.

SavchenkoValeriy commented:
> > > > Alive2 struggles (times out) with 32 bit vectors for this particular proof: https://alive2.llvm.org/ce/z/S-78k- In my z3 proof, even 16 bit took quite some times.
> > > 
> > > 
> > > Please confirm if I shrunk the inputs correctly: https://alive2.llvm.org/ce/z/AABL4v (I guess you missed poison-generating flags in the source, as otherwise your target function is defined for a less pairs of inputs than the source.)
> > 
> > 
> > In your example, we can always infer that the result is < 65535 just because we are extending i8 to i32. The correct shrinking would include extending N / 4 to size N, but actually we can infer that the result is < 2^(N/2). Like this: https://alive2.llvm.org/ce/z/5A5mCg
> 
> Oh, you're right, missed that the sum fits N/2; so inferring nsw/nuw should make sense as the result will never overflow: https://alive2.llvm.org/ce/z/_rAdQg.

Exactly!

dtcxzyw commented on llvm/lib/Analysis/ValueTracking.cpp:
```diff
@@ -369,6 +503,15 @@ static void computeKnownBitsAddSub(bool Add, const Value *Op0, const Value *Op1,
+  if (Add) {
+    // Try to match lerp pattern and combine results
+    const KnownBits LerpKnown =
+        computeKnownBitsFromLerpPattern(Op0, Op1, DemandedElts, Q, Depth);
```
Pass `KnownOut` into the function and modify the result in the function.

dtcxzyw commented on llvm/lib/Analysis/ValueTracking.cpp:
```diff
@@ -350,6 +350,140 @@ unsigned llvm::ComputeMaxSignificantBits(const Value *V, const DataLayout &DL,
+// graphics in the shape of alpha blending.
+//
+// Returns unknown bits if the pattern doesn't match or constraints don't apply
+// to the given operands.
```
Use `///` for header comments.

dtcxzyw commented on llvm/lib/Analysis/ValueTracking.cpp:
```diff
@@ -350,6 +350,140 @@ unsigned llvm::ComputeMaxSignificantBits(const Value *V, const DataLayout &DL,
+    // that the constant will be the RHS. xor of two positive integers is
+    // guaranteed to be non-negative as well.
+    return m_CombineOr(m_NUWSub(m_Value(B), m_Value(C)),
+                       m_Xor(m_Value(C), m_Value(B)));
```
Missing check for the constraint `C <=u B`.

dtcxzyw replied:
> For xor, this comes out of the box since xor doesn't wrap.

But you cannot refine `xor c, b` into `sub nuw c, b` by assuming that `xor c, b` is introduced as the canonical form of `sub nuw Mask, b`. A counterexample: https://alive2.llvm.org/ce/z/sWwAPJ

dtcxzyw reviewed (commented):
Can you please provide a generalized alive2 proof?
See https://llvm.org/docs/InstCombineContributorGuide.html#proofs.

SavchenkoValeriy commented:
> > > Can you please provide a generalized alive2 proof? See https://llvm.org/docs/InstCombineContributorGuide.html#proofs.
> > 
> > 
> > I believe I answered it here: [#166378 (comment)](https://github.com/llvm/llvm-project/pull/166378#issuecomment-3487318866)
> > In short, I'm not improving any particular transformation but an analysis and analysis can't be proven via alive2. I provide a general z3 proof for the exact logic that I encoded. Please tell me if you have any questions about that proof.
> 
> z3 proof is ok. But you have to handle poison stuff yourself (e.g., `sub nuw`) to ensure your code matches the proof. It is also a burden for reviewers to check your proof. With Alive2 we can also check if additional poison-generating flags on adds/muls provide a better estimation of knownbits.

With the Alive2 proof it's also not fully true again because I need to prove the analysis and not the transformation.
In any case, something like this can be used for `sub nuw` https://alive2.llvm.org/ce/z/5egb4W and for `xor` https://alive2.llvm.org/ce/z/F45enJ proving the maximum property. Leading zeros property of unsigned maximum is, however, trivial enough. Please tell me if that works for you and is enough.

dtcxzyw commented on llvm/lib/Analysis/ValueTracking.cpp:
```diff
@@ -350,6 +350,152 @@ unsigned llvm::ComputeMaxSignificantBits(const Value *V, const DataLayout &DL,
+    // for constant b's). Even though xor is symmetrical, canonicalization
+    // ensures that the constant will be the RHS. We have additional checks
+    // later on to ensure that this xor operation is equivalent to subtraction.
+    return m_CombineAnd(m_BinOp(SubBC),
```
We have an overload `m_Instruction(Instruction *&I, const MatchTy &Match)` for this purpose.

dtcxzyw reviewed (approved):
LGTM. Thanks!
